### DISCUSSION TITLE
DataGrid Dynamic Column Binding

### DISCUSSION THREAD

**Discussion by JBrolin90** [AvaloniaUI/Avalonia#17534] [Q&A] [answered]
### Intro
Hello everyone!

I have this class that wraps a System.Data.DataRow class to facilitate binding to Avalonia UI Elements. However, when I bind, using index syntax, the cells does not listen to property changed events. I have made a small Avalonia.App which isolates the issue. It is here: https://github.com/JBrolin90/DataTableGrid

Would be wonderful if someone could help a shed some light on this my issue.

### Breakdown
This call wraps System,Data.DataRow:
`public class ObservableDataRow : INotifyPropertyChanged`

In this class, I have two Access properties, "Name" and "Age" like so. When I bind these properties to my DataGrid, everything works as expected. 
```
    public string Name
    {
        get => Row[0].ToString() ?? "No Name"; // called by DataGrid during setup
        set => Row[0] = value;
    }

```
And 
```
   public string Age
    {
        get => Row[1].ToString() ?? "-1"; // called by DataGrid during setup
        set => Row[1] = value;
    }

```
Both of them use this indexer:
```
    public object this[int index]
    {
        get => Row[index]; // called by DataGrid during setup
        set => Row[index] = value;
    }

```
This is how I route the events:
```
    private void Table_ColumnChanged(object sender, DataColumnChangeEventArgs e)
    {
        if (e.Row == row && e.Column != null) // e.Column is null when the row is deleted
        {
            string columnName = e.Column.ColumnName;
            int ordinal = e.Column.Ordinal;
            OnPropertyChanged($"{columnName}");
            OnPropertyChanged($"[{columnName}]");
            OnPropertyChanged($"[{ordinal}]");
        }
    }

```
### Issue

What I really want to do is to bind using the Row[index] without having to define named properties.

When I Bind like this

```
        //Two way binding works as expected for the following two bindings 
        dataGrid.Columns.Add(new DataGridTextColumn { Header = "Name", Binding = new Binding("Name") });
        dataGrid.Columns.Add(new DataGridTextColumn { Header = "Age", Binding = new Binding("Age") });

```
But with these bindings, the grid does not listen to property change events

```
        //For the following two bindings, the texboxes updates the underlying data source, but the texboxes do not respond to Property changed events.
        //In addition, I have to set IsReadOnly to false to enable editing.
        dataGrid.Columns.Add(new DataGridTextColumn { Header = "Name", Binding = new Binding("[0]"), IsReadOnly = false});
        dataGrid.Columns.Add(new DataGridTextColumn { Header = "Age", Binding = new Binding("[1]"), IsReadOnly = false });
 
```

### Background

I'm trying to wrap my head around the datagrid and it's bindings. For that reason I implemented my own version of  ObservableCollection<T> from scratch and I made work, so that is fine. However, my goal is to make a (set of) classes to enable easy connection of a DataTable to the DataGrid. Very briefly, I have implemented something that makes sense to me but I have a few problems. The first is that the DataGrid does not ask for IsReadonly for my DataTable adapter but it does for my ObservableCollection class. I cannot really find any obvious way to debug what is going on.

Thank you for reading all the way here

**Comment by timunie** [comment]
You're missing OnPropertyChanged("Item"');, see https://github.com/AvaloniaUI/avalonia-docs/issues/51

It comes unfortunately with it's own drawbacks.

**Reply by JBrolin90** [reply]
OK, let me go ahead and try that out

Thank you @timunie for very quick response !

**Comment by timunie** [comment, marked as answer]
I've another idea. Your Indexer could return a CellViewModel:

```cs
public class CellViewModel : INotifyPropertyChanged
{
   // Pseudo code. You need to impl INPC correctly here
   public object? Value { get; set; }
}
```

And your Binding:
```cs
new Binding("[0].Value");
```

**Reply by JBrolin90** [reply]
That is an interesting idea, wrapping the object in an attempt to make it easier for the parser, right?! I'll try that and get back here.

_(Just to mention, the OnPropertyChanged("item") did not work immediately but I still need to check that I did it the correct way)_

**Reply by JBrolin90** [reply]
Your idea with the CellViewModel class did the trick. It now work as expected. This is how I implemented the CellViewModel (I ended up calling it ObservableDataCell) 

using System.ComponentModel;
using System.Data;

```
namespace DataTableGrid.ViewModels.TableBase;

internal class ObservableDataCell : INotifyPropertyChanged
{
    public event PropertyChangedEventHandler? PropertyChanged;

    private readonly DataRow row;
    private readonly DataColumn column;

    public ObservableDataCell(DataRow row, DataColumn column)
    {
        this.column = column;
        this.row = row;
        row.Table.ColumnChanged += Table_ColumnChanged;
    }

    private void Table_ColumnChanged(object sender, DataColumnChangeEventArgs e)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Value)));
    }
    public object? Value
    {
        get => row[column];
        set { if (row[column] != value) row[column] = value; }
    }
}

```

And now I can easily setup a "DataTableGrid" dynmically like this:

```
    private void BuildDataGridColumns()
    {
        dataGrid.Columns.Clear();
        foreach (DataColumn column in observableDataTable.Columns)
        {
            DataGridTextColumn textColumn = new()
            {
                Header = column.Caption ?? column.ColumnName,
                Binding = new Binding($"[{column.Ordinal}].Value"),
                IsReadOnly = false,
            };
            dataGrid.Columns.Add(textColumn);
        }
    }

```

You'll find the full solution here https://github.com/JBrolin90/DataTableGrid

Thanks @timunie
